src/pybind/mgr/restful/module.py

```diff
@@ -0,0 +1,511 @@
+            ),
+            'waiting': map(
+                lambda x: (x.command, x.outs, x.outb),
+                self.waiting
+            ),
+            'failed': map(
+                lambda x: (x.command, x.outs, x.outb),
+                self.failed
+            ),
+            'is_waiting': self.is_waiting(),
+            'is_finished': self.is_finished(),
+            'has_failed': self.has_failed(),
+            'state': self.get_state(),
+        }
+
+
+
+class Module(MgrModule):
+    COMMANDS = []
+
+    def __init__(self, *args, **kwargs):
+        super(Module, self).__init__(*args, **kwargs)
+        global instance
+        instance = self
+
+        self.requests = []
+        self.requests_lock = threading.RLock()
+
+        self.tokens = {}
+        self.disable_auth = False
+
+        self.shutdown_key = str(uuid4())
+
+        self.app = None
+        self.api = None
+
+
+    def serve(self):
+        try:
+            self._serve()
+        except:
+            self.log.error(str(traceback.format_exc()))
+
+
+    def _serve(self):
+        # Load stored authentication tokens
+        self.tokens = self.get_config_json("tokens") or {}
+
+        # Create the Flask app
+        self.app = Flask('restful')
+        self.app.config['RESTFUL_JSON'] = {
+            'sort_keys': True,
+            'indent': 4,
+            'separators': (',', ': '),
+        }
+        self.api = Api(self.app)
+
+        # Add the resources as defined in api module
+        for _obj in dir(api):
+            obj = getattr(api, _obj)
+            # We need this try statement because some objects (request)
+            # throw exception on any out of context object access
+            try:
+                _endpoint = getattr(obj, '_endpoint', None)
+            except:
+                _endpoint = None
+            if _endpoint:
+                self.api.add_resource(obj, _endpoint)
+
+        self.cert = self.get_config_json('cert')
+        self.pkey = self.get_config_json('pkey')
+
+        # Create a new shared cert if it does not exist, yet
+        if not self.cert or not self.pkey:
+            (self.cert, self.pkey) =  self.create_cert()
+            self.set_config_json('cert', self.cert)
+            self.set_config_json('pkey', self.pkey)
+
+        # use SSL context for https
+        context = SSL.Context(SSL.TLSv1_METHOD)
+        context.use_certificate(
+            crypto.load_certificate(crypto.FILETYPE_PEM, self.cert)
+        )
+        context.use_privatekey(
+            crypto.load_privatekey(crypto.FILETYPE_PEM, self.pkey)
+        )
+
+        self.app.run(host='0.0.0.0', port=8002, ssl_context=context)
+
+
+    def shutdown(self):
+        try:
+            self._shutdown()
+        except:
+            self.log.error(str(traceback.format_exc()))
+
+
+    def _shutdown(self):
+        # We need request module to get the shutdown function
+        requests.post(
+            'https://localhost:8002/shutdown',
```

Comment:
Does the werkzeug server really not have a better way to shut down than making a process POST to itself?